overlord/snapstate/handlers.go

```diff
@@ -2247,7 +2250,7 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 
 	// links the new revision to current and ensures a shared base prefix
 	// directory for parallel installed snaps
-	rebootInfo, err := m.backend.LinkSnap(newInfo, deviceCtx, linkCtx, perfTimings)
+	err = m.backend.LinkSnap(newInfo, deviceCtx, linkCtx, perfTimings)
 	// defer a cleanup helper which will unlink the snap if anything fails after
 	// this point
 	defer func() {
```

Comment:
is the reboot setup state tested for when we hit this defer? I suppose in this case we shouldn't reboot?

Reply:
This is a good question. There is no revert of the restart set-up for next boot in `backend.UnlinkSnap`, and I actually wondered how undoing this worked in case of an error before rebooting. Tbh it looks like a bug that we are not tackling this currently, unless I am wrong and it is done somewhere. In fact, the proposed change can help here as we could possibly move the call to `backend.MaybeSetNextBoot` to `snapstate.FinishTaskWithRestart`, right at the end, so setting up bootloader/modeenv is done last and there can be no errors after it and before reboot.

Reply:
I've added a TODO now, I think that the bug that I describe is possible but difficult to trigger because after the link backend calls most of the code just handles internal state and writes files. There is the possibility of a forced restart while in the middle of that code though. To minimize this I have put the call to `MaybeSetNextBoot` a bit later.

I have now also added a test forcing a failure in between.